### ISSUE TITLE
Feature Request: Locally Connected Layer

### ISSUE THREAD

**Issue by ronrest** [pytorch/pytorch#499]
The Lua version of Torch implements a [Locally Connected Layer](https://github.com/torch/nn/blob/master/doc/convolution.md#spatialconvolutio) (Convolution without shared weights). It would be good to have an implementation of this available on the python API version as well.

**Comment by 1zb** [comment]
I came up with the same issue and found `torch.nn.backends.thnn.backend.SpatialConvolutionLocal`.

For anyone who want to use locally connected layer in PyTorch,

```python
inputs=Variable(torch.randn(1,3,32,32)) # N x inC x inH x inW
weight=Variable(torch.randn(30,30,64,3,3,3)) # outH x outW x outC x inC x kH x kW
bias=Variable(torch.randn(64,30,30)) # outC x outH x outW

## arguments: kH, kW, strideH, strideW, padH, padW, inH, inW, outH, outW
local=torch.nn.backends.thnn.backend.SpatialConvolutionLocal(3, 3, 1, 1, 0, 0, 32, 32, 30, 30)
outputs=local(inputs, weight, bias) # N x outC x outH x outW
```

**Comment by apaszke** [comment]
This should be easy to add once #1048 is implemented

**Comment by apaszke** [comment]
Oh, if the auto-generated function works then we can expose it. @1zb can you send a PR that creates a module to do it? You can base the implementation on other `Conv*` modules, and use the function you found in `forward`

**Comment by 1zb** [comment]
Sure, I am happy to do this.

**Comment by ibulu** [comment]
I would like to request the same for 1D data, Conv1DLocal?

**Comment by bwesen** [comment]
Beware that the torch.nn.backends.thnn.backend.SpatialConvolutionLocal function mentioned above is very slow, feeding it batches you get no parallelisation (minibatch of 10 is 10x the time of a 1 batch). At least in pyTorch 0.3.1. Other high-level methods using Unfold seem to work better.

**Comment by kausic94** [comment]
@1zb 
I try this, but it say that required arguement bias is missing despite it being present. Do you know what the problem could be?

```
## arguments: kH, kW, strideH, strideW, padH, padW, inH, inW, outH, outW
local=torch.nn.backends.thnn.backend.SpatialConvolutionLocal(3, 3, 1, 1, 0, 0, 32, 32, 30, 30)
outputs=local(inputs, weight, bias) # N x outC x outH x outW
```

**Comment by bwesen** [comment]
I was also hit by this breakage, and ended up re-learning CUDA and wrote the kernels I needed myself and added them as pyTorch extensions. This turned out to be far less work than I spent trying to figure out how to mash the other tensor functions together to do the same thing and it was 10-20x faster as well than those functions even while they worked (the im2col-based unfold things). Mainly due to the massive memory usage of the latter (it has to keep an intermediary gigantonourmous array in RAM).

But for a general case, building up functions to work in a library with multiple frontends and backends, it's much more work than for the very specific fixed kernels I needed. I guess my point being, if you have a specific case that doesn't need flexibility, adding a cuda extension is not as hard as it might seem.

**Comment by shir994** [comment]
I was playing around with PyTorch unfold operation. If I understand correctly, LocalConv2d could be done, if we first use torch.nn.Unfold, and then apply kernel with different weights to each of the last dimension. I think, it could be implemented as follows, as very ad-hoc approach:

```
class LocallyConnected2d(nn.Module):
    def calculate_spatial_output_shape(self, input_shape, kernel_size, dilation, padding, stride):
        return [np.floor((input_shape[index] + 2 * padding[index] - dilation[index] * (kernel_size[index] - 1) - 1) /
                stride[index] + 1).astype(int) for index in range(len(input_shape))]
    
    def __init__(self, input_shape, in_channels, out_channels, kernel_size, dilation=(1,1), padding=(0,0), stride=(1,1)):
        super().__init__()
        self.kernel_size = kernel_size
        self.out_channels = out_channels
        self.dilation = dilation
        self.padding = padding
        self.stride = stride
        
        self.output_height, self.output_width = self.calculate_spatial_output_shape(input_shape, kernel_size,
                                                                               dilation, padding, stride)
        self.weight_tensor_depth = in_channels * kernel_size[0] * kernel_size[1]
        self.spatial_blocks_size = self.output_height * self.output_width
        self.weights = nn.Parameter(torch.empty((1, self.weight_tensor_depth, self.spatial_blocks_size, out_channels),
                                   requires_grad=True, dtype=torch.float))
        self.bias = nn.Parameter(torch.empty((1, out_channels, self.output_height, self.output_width),
                                requires_grad=True, dtype=torch.float))
        
        torch.nn.init.xavier_uniform_(self.weights)
        torch.nn.init.xavier_uniform_(self.bias)
        
    def forward(self, input):
        input_unf = torch.nn.functional.unfold(input, self.kernel_size, dilation=self.dilation,
                                               padding=self.padding, stride=self.stride)
        local_conv_unf = (input_unf.view((*input_unf.shape, 1)) * self.weights)
        return local_conv_unf.sum(dim=1).transpose(2, 1).reshape(
                                        (-1, self.out_channels, self.output_height, self.output_width)) + self.bias
```
Although, I am very suspicious about speed performance of such implementation, since I do no use any of raw PyTorch backend methods.

**Comment by bwesen** [comment]
My conv2d_local function I used before rewriting it in custom CUDA, if anyone is interested. It's actually from an unmerged PR somewhere else here.

Note that it's *extremely* slow and memory-inefficient to use im2col. It's surely easy and flexible, but it requires the GPU to unpack all the overlapping kernel possibilities in the input array to an intermediary storage and then apply the weight matrix in that space. Doing it "in place" in RAM is about 100x faster (yes, I benchmarked it). For small arrays I'm sure there's much less difference though.

```
# This is from this unmerged PR:
# https://github.com/pytorch/pytorch/pull/1583

def conv2d_local(input, weight, bias=None, padding=0, stride=1, dilation=1):
    if input.dim() != 4:
        raise NotImplementedError("Input Error: Only 4D input Tensors supported (got {}D)".format(input.dim()))
    if weight.dim() != 6:
        # outH x outW x outC x inC x kH x kW
        raise NotImplementedError("Input Error: Only 6D weight Tensors supported (got {}D)".format(weight.dim()))

    outH, outW, outC, inC, kH, kW = weight.size()
    kernel_size = (kH, kW)

    # TODO: this requires a huge amount of temporary RAM, for the unfold and matmul.
    
    # N x [inC * kH * kW] x [outH * outW]
    cols = torch.nn.functional.unfold(input, kernel_size, dilation=dilation, padding=padding, stride=stride)
    cols = cols.view(cols.size(0), cols.size(1), cols.size(2), 1).permute(0, 2, 3, 1)
    
    out = torch.matmul(cols, weight.view(outH * outW, outC, inC * kH * kW).permute(0, 2, 1))
    out = out.view(cols.size(0), outH, outW, outC).permute(0, 3, 1, 2)
    
    if bias is not None:
        out = out + bias.expand_as(out)
    return out
```

**Comment by kakaxi314** [comment]
@bwesen I tried to unpack the feature and then did matmul to implement the locally connected layer. I worry more about GPU memory than speed. I found my python implementation costs large sums of GPU memory. As I have never wrote CUDA code, I'm interested whether your custom CUDA conv2d_local function would occupy less GPU memory. And also, would you like to share the CUDA version code?

**Comment by Blusea** [comment]
```python
class LocalLinear(nn.Module):
    def __init__(self,size,out_features,stride=1,bias=True):
        super(LocalLinear, self).__init__()
        self.lc = nn.Linear(size,out_features,bias)
        self.size = size
        self.step = stride

    def forward(self, x:torch.Tensor):
        batch_size = x.shape[0]
        x = x.unfold(0,self.size,self.step)
        x = self.lc(x)
        return torch.reshape(x,(batch_size,-1))
```

**Comment by Blusea** [comment]
```python
import torch,torch.nn as nn,torch.nn.functional as F
import copy
class LocalLinear(nn.Module):
    def __init__(self,in_features,out_features,kernel_size,stride=1,bias=True):
        super(LocalLinear, self).__init__()
        self.kernel_size = kernel_size
        self.stride = stride

        in_features = (in_features-self.kernel_size)//self.stride+1
        self.lc = nn.ModuleList([copy.deepcopy(nn.Linear(kernel_size,out_features,bias=bias))
                                 for _ in range(in_features)])

    def forward(self, x:torch.Tensor):
        x = x.unfold(-1,size=self.kernel_size,step=self.stride)
        fold_num = x.shape[1]
        x= torch.cat([self.lc[i](x[:,i,:]) for i in range(fold_num)],1)
        return x
```

**Comment by Blusea** [comment]
```python
import torch,torch.nn as nn,torch.nn.functional as F
import copy
class LocalLinear(nn.Module):
    def __init__(self,in_features,out_features,kernel_size,stride=1,bias=True):
        super(LocalLinear, self).__init__()
        self.kernel_size = kernel_size
        self.stride = stride

        fold_num = (in_features-self.kernel_size)//self.stride+1
        self.lc = nn.ModuleList([copy.deepcopy(nn.Linear(kernel_size,out_features,bias=bias))
                                 for _ in range(fold_num)])

    def forward(self, x:torch.Tensor):
        x = x.unfold(-1,size=self.kernel_size,step=self.stride)
        fold_num = x.shape[1]
        x= torch.cat([self.lc[i](x[:,i,:]) for i in range(fold_num)],1)
        return x
```

**Comment by Blusea** [comment]
```python
import torch,torch.nn as nn,torch.nn.functional as F
from copy import deepcopy
class LocalLinear(nn.Module):
    def __init__(self,in_features,local_features,kernel_size,stride=1,bias=True):
        super(LocalLinear, self).__init__()
        self.kernel_size = kernel_size
        self.stride = stride

        fold_num = (in_features-self.kernel_size)//self.stride+1
        self.lc = nn.ModuleList([deepcopy(nn.Linear(kernel_size,local_features,bias=bias))
                                 for _ in range(fold_num)])

    def forward(self, x:torch.Tensor):
        x = x.unfold(-1,size=self.kernel_size,step=self.stride)
        fold_num = x.shape[1]
        x= torch.cat([self.lc[i](x[:,i,:]) for i in range(fold_num)],1)
        return x

```

**Comment by Blusea** [comment]
```python
class LocalLinear(nn.Module):
    def __init__(self,in_features,local_features,kernel_size,padding=0,stride=1,bias=True):
        super(LocalLinear, self).__init__()
        self.kernel_size = kernel_size
        self.stride = stride
        self.padding = padding

        fold_num = (in_features+2*padding-self.kernel_size)//self.stride+1
        self.weight = nn.Parameter(torch.randn(fold_num,kernel_size,local_features))
        self.bias = nn.Parameter(torch.randn(fold_num,local_features)) if bias else None

    def forward(self, x:torch.Tensor):
        x = F.pad(x,[self.padding]*2,value=0)
        x = x.unfold(-1,size=self.kernel_size,step=self.stride)
        x = torch.matmul(x.unsqueeze(2),self.weight).squeeze(2)+self.bias
        return x
```

**Comment by mru4913** [comment]
@shir994  
Your padding is not being used.

**Comment by mehameha998** [comment]
> ```python
> class LocalLinear(nn.Module):
>     def __init__(self,in_features,local_features,kernel_size,padding=0,stride=1,bias=True):
>         super(LocalLinear, self).__init__()
>         self.kernel_size = kernel_size
>         self.stride = stride
>         self.padding = padding
> 
>         fold_num = (in_features+2*padding-self.kernel_size)//self.stride+1
>         self.weight = nn.Parameter(torch.randn(fold_num,kernel_size,local_features))
>         self.bias = nn.Parameter(torch.randn(fold_num,local_features)) if bias else None
> 
>     def forward(self, x:torch.Tensor):
>         x = F.pad(x,[self.padding]*2,value=0)
>         x = x.unfold(-1,size=self.kernel_size,step=self.stride)
>         x = torch.matmul(x.unsqueeze(2),self.weight).squeeze(2)+self.bias
>         return x
> ```

Have you checked this implementation? @sheljoy

**Comment by Ajk4** [comment]
@bwesen 

Hi, is there any chance you could contribute your CUDA implementation to the community?

**Comment by ParadoxRobotics** [comment]
Hi,
Is it possible to have also the transposed version of a locally connected network ?
It seems the fold operation doesn't support 4D tensor like the fold operation used here...

**Comment by KushajveerSingh** [comment]
@apaszke @soumith  I can work on this issue. Can you give me some pointer on how should I proceed?

**Comment by yuzhms** [comment]
Any progress in 2021?

**Comment by grau4** [comment]
Any progress in Dec-2022?

**Comment by yw5aj** [comment]
Any helping hand would be appreciated... Jan 2023 :(

**Comment by DanielWicz** [comment]
Waiting :-)

**Comment by sheljoy** [comment]
@gsygsy96  checked and updated  with  [ConvLocal](https://github.com/google/flax/blob/main/tests/linen/linen_linear_test.py) in Jax,  of which test case in flax.

``` python
import math,numpy as np
import torch
import torch.nn as nn

import math,numpy as np
import torch
import torch.nn as nn

class LocalLinear(nn.Module):

    def __init__(self, in_features, out_features: int, kernel_size,
                 dilation=1, padding=0, stride=1, bias: bool = True) -> None:

        super(LocalLinear, self).__init__()

        self.in_features = in_features
        self.out_features = out_features

        self.kernel_size = kernel_size
        self.kernel_area = int(np.prod(self.kernel_size))
        self.dilation = dilation
        self.padding = padding
        self.stride = stride

        unfold_features = np.prod(self.get_patch_size())
        self.unfold = nn.Unfold(kernel_size, dilation, padding, stride)
        self.weight = nn.Parameter(torch.empty(int(out_features), self.kernel_area, int(unfold_features)))
        if bias:
            self.bias = nn.Parameter(torch.empty(self.kernel_area, int(unfold_features)))
        else:
            self.register_parameter('bias', None)

    def reset_parameters(self):
        nn.init.kaiming_uniform_(self.weight, a=math.sqrt(5))
        if self.bias is not None:
            fan_in, _ = nn.init._calculate_fan_in_and_fan_out(self.weight)
            bound = 1 / math.sqrt(fan_in) if fan_in > 0 else 0
            nn.init.uniform_(self.bias, -bound, bound)

    def get_patch_size(self):
        _, *spatial_size = self.in_features

        return [
            np.floor((spatial_size[d]+2*self.padding-self.dilation*(self.kernel_size[d]-1)-1)/self.stride+1)
            for d in range(len(kernel_size))
        ]

    def forward(self, input: torch.Tensor):

        input_unf = self.unfold(input)
        batch_size, _, L = input_unf.shape
        input_unf = input_unf.reshape(batch_size, -1, self.kernel_area, L)
        output = torch.einsum("nckl,okl->nokl",input_unf, self.weight)
        if self.bias is not None:
            output = output + self.bias
        return output.sum(-2)

```

**Comment by bramgrooten** [comment]
I would love to see this as well!

**Comment by YonatanSimson** [comment]
Any news?

**Comment by dxiaalign** [comment]
Any progress in Nov-2023?

**Comment by whubaichuan** [comment]
Any progress in April-2024? 

It seems SpatialConvolutionLocal is not available in PyTorch now. @ronrest @apaszke

**Comment by XichiShen** [comment]
+1
